pageserver/src/tenant/timeline.rs

```diff
@@ -3133,6 +3045,7 @@ impl Timeline {
 
         let (sender, receiver) = tokio::sync::oneshot::channel();
         // Spawn a task so that download does not outlive timeline when we detach tenant / delete timeline.
+        let self_clone = self.myself.upgrade().expect("timeline is gone");
```

Comment:
Might be `.context` here instead? 
It's not obvious from the code, that `download_remote_layer` can never be called after the timeline drop, so better cancel fast, instead of panicking here.

Reply:
This is a "can't happen" case. If you have a reference to `self`, then the Timeline hasn't been dropped, and `self.myself.upgrade()` must work. In fact maybe we should add a function to hide the `expect`, something like:
```
fn myself(&self) -> Arc<Self> {
    // `self.myself` is a Weak reference to the Arc of 'self`, so as long as the Timeline is live, `self.myself.upgrade()`
    // cannot fail.
    self.myself.upgrade.expect("timeline is gone")
}
```

Reply:
I have a great PR on the similar topic: https://github.com/neondatabase/neon/issues/3252

It concentrates all weak ref upgrades in the single place with some context.
Reviews are welcome.